### ISSUE TITLE
Building with -O0 causes programs/zstd to fail to link

### ISSUE THREAD

**Issue by eli-schwartz** [facebook/zstd#2950]
For example, configuring the meson build using

```
$ meson setup --buildtype=debug builddir/
$ ninja -C builddir/
[50/50] Linking target programs/zstd
FAILED: programs/zstd 
cc  -o programs/zstd programs/zstd.p/.._.._.._programs_zstdcli.c.o programs/zstd.p/.._.._.._programs_util.c.o programs/zstd.p/.._.._.._programs_timefn.c.o programs/zstd.p/.._.._.._programs_fileio.c.o programs/zstd.p/.._.._.._programs_benchfn.c.o programs/zstd.p/.._.._.._programs_benchzstd.c.o programs/zstd.p/.._.._.._programs_datagen.c.o programs/zstd.p/.._.._.._programs_dibio.c.o programs/zstd.p/.._.._.._programs_zstdcli_trace.c.o programs/zstd.p/.._.._.._lib_common_xxhash.c.o -Wl,--as-needed -Wl,--no-undefined '-Wl,-rpath,$ORIGIN/../lib' -Wl,-rpath-link,/tmp/zstd-1.5.1/build/meson/builddir/lib -Wl,--start-group lib/libzstd.so.1.5.1 /usr/lib/libz.so /usr/lib/liblzma.so /usr/lib/liblz4.so -Wl,--end-group
/usr/bin/ld: programs/zstd.p/.._.._.._programs_dibio.c.o: in function `ERR_getErrorName':
/tmp/zstd-1.5.1/build/meson/builddir/../../../programs/../lib/common/error_private.h:75: undefined reference to `ERR_getErrorString'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

I can reproduce the failure by setting the buildtype=plain, and make the error go away by additionally adding `CFLAGS=-O1`.

Although presumably zstd may rather suffer for building without any optimizations, undefined symbols seems odd.

**Comment by terrelln** [comment]
Thats super odd, and definitely unexpected. It sounds like the meson build may be missing an object file.

Are you still having this issue?

**Comment by eli-schwartz** [comment]
I am not at my computer right now so I can't test again, but 2 weeks ago when I reported this issue it was present in the 1.5.1 stable tag and I'm 99% sure I verified it also on what was at the time the latest commit on `dev`. I doubt anything would have changed since then.

I would be surprised if it was due to missing object files as surely the shared library would fail too.

It might have something to do with `-fvisibility`? I did contribute a couple changes to the meson.build files along those lines, and if I recall correctly it was basically impossible to link the test programs to libzstd.so when making that change, as they poked at private symbols... What confused the heck out of me is that using optimizations *solves* the problem rather than optimizing symbols away as unneeded and *causing* a problem.

**Comment by terrelln** [comment]
I've reproduced the issue in dev. I'm not sure what is going on yet, but we'll fix it by the next release.

**Comment by terrelln** [comment]
Ok, the reason it fails in debug mode but not opt mode is that the function that needs `ERR_getErrorName()` is unused. So in opt mode the function is removed from the executable.

The fix is simple, remove the unused `#include "../lib/common/error_private.h"` from `programs/dibio.c`.

The reason we didn't catch this is we don't really support linking the CLI against a shared library. Since it relies on the version matching exactly, and we rely on some symbols that aren't exported by the shared library.

meson should really be linking the CLI against the static library.

**Comment by eli-schwartz** [comment]
> The reason we didn't catch this is we don't really support linking the CLI against a shared library. Since it relies on the version matching exactly

I don't think this is a good reason. ;) They are both distributed in the same repository, built by the same build system, and in most distros contained in the same package -- and in all other distros, contained in shared subpackages of a single source package (and bound to the same rebuild number). Hence, the version will always match exactly *anyway*, unless something very strange happened.

Meanwhile, for a 2.8 mb package over a third of that is just the additional megabyte caused by the zstd tool being statically linked:
```
$ du -sh builddir*/programs/zstd
196K	builddir/programs/zstd
1.2M	builddir-static/programs/zstd
```

I'm not saying one mb is the end of the world, but... what are the advantages it gets out of that? Does compression/decompression happen faster because the zstd tool avoids opening a shared library? I suspect not. It seems more like a waste. And there are surely people out there whose use cases value reducing the storage footprint, even if those use cases probably aren't "the average user's average desktop system".

People who want a statically linked version can already get it if they build with meson using `-Ddefault_library=static`.

> and we rely on some symbols that aren't exported by the shared library.

We do this too in a different project I'm involved in. We handled that by building a library called "libcommon.a", which is not installed. It's linked to by both the shared library (or its objects included in the static library, of course), as well as by the CLI tool.

The purpose of this "libcommon" is to include code that for one reason or another is useful in our own programs, but isn't part of the shared library public ABI. Instead of declaring that we cannot link the tool against the shared library at all, we just keep separate accounting of all private functions that have multiple consumers. And yes, those functions do need to be accounted for because in order to be available for both uses, they need to live in separate files.

If that sounds like a useful idea, feel free to do that in zstd too. :)

**Comment by terrelln** [comment]
We can definitely remove the include that breaks this build, and fix it for 1.5.2.

But, for now, linking the CLI against the shared library isn't truly supported. It'd probably work, as you see now in opt builds, since I think all the symbols we use are `static`.

I'm open for discussion about supporting linking the CLI against a shared library that is built against the same commit of zstd. And ensuring that we have good test coverage for that scenario. But that is a medium-long term topic.

Even longer term is supporting linking against any version of the library. That would require a significant refactoring of the CLI, as we rely on a lot of unstable features. But it is technically possible.

**Comment by eli-schwartz** [comment]
I opened a new issue to discuss that topic, as it's out of scope for this ticket which talks about a build failure. :)
